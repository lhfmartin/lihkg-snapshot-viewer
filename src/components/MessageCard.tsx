import Message from '../interfaces/message';
import ThumbDownOffAltIcon from '@mui/icons-material/ThumbDownOffAlt';
import ThumbUpOffAltIcon from '@mui/icons-material/ThumbUpOffAlt';
import styles from './MessageCard.module.css';
import moment from 'moment';
import Stack from '@mui/material/Stack';
import { useContext } from 'react';
import { PushClickedQuoteFromMsgNumsContext } from './ThreadViewer';

export default function MessageCard({
  message,
  originalPosterUserId,
}: {
  message: Message;
  originalPosterUserId: Message['user']['user_id'];
}) {
  const pushClickedQuoteFromMsgNums = useContext(
    PushClickedQuoteFromMsgNumsContext,
  );
  return message.status > 1 ? (
    <></>
  ) : (
    <div id={message.msg_num} className={styles.card}>
      <div className={styles.metadata}>
        <span
          className={
            message.user.user_id === originalPosterUserId
              ? styles.msgNumOp
              : styles.msgNum
          }
        >
          #{message.msg_num}
        </span>
        <span
          className={
            styles[
              message.user.level === '999'
                ? 'username--admin'
                : `username--${message.user_gender}`
            ]
          }
        >
          {message.user_nickname}
        </span>
        <span className={styles.replyTime}>
          {moment(new Date(message.reply_time * 1000)).format(
            'YYYY年M月D日 HH:mm:ss',
          )}
        </span>
      </div>
      {message.quote && (
        <blockquote className={styles.quote}>
          <a
            onClick={
              message.quote!.status > 1
                ? undefined
                : () => {
                    pushClickedQuoteFromMsgNums(
                      message.msg_num,
                      message.quote!.msg_num,
                    );
                  }
            }
            href={
              message.quote!.status > 1
                ? undefined
                : `#${message.quote.msg_num}`
            }
          >{`Go to quote (#${message.quote.msg_num})`}</a>
        </blockquote>
      )}
      <div
        className={styles.msg}
        dangerouslySetInnerHTML={{ __html: message.msg }}
      />
      <div className={styles.votes}>
        <Stack
          direction='row'
          alignItems='stretch'
          justifyContent='space-around'
          gap={2}
        >
          <Stack direction='row' alignItems='center' gap={1}>
            <ThumbUpOffAltIcon fontSize='small' /> {message.like_count}
          </Stack>
          <Stack direction='row' alignItems='center' gap={1}>
            <ThumbDownOffAltIcon fontSize='small' /> {message.dislike_count}
          </Stack>
        </Stack>
      </div>
    </div>
  );
}
