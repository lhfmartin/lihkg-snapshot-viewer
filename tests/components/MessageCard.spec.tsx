import { expect, it } from 'vitest';
import { render } from '@testing-library/react';
import MessageCard from '@/src/components/MessageCard';
import Message from '@/src/interfaces/message';

it('renders MessageCard', async () => {
  const message: Message = {
    msg: 'This is a test message',
    msg_num: '1',
    user_gender: 'M',
    user_nickname: 'TestUser',
    user: {
      user_id: '2',
      level: '10',
    },
    reply_time: 1625079600,
    like_count: '10',
    dislike_count: '2',
    status: 1,
  };

  const { container } = await render(
    <MessageCard message={message} originalPosterUserId='1' />,
  );
  expect(container).toMatchSnapshot();
});
