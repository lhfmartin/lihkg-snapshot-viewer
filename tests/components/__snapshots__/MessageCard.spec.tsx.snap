// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders MessageCard 1`] = `
<div>
  <div
    class="_card_0a2670"
    id="1"
  >
    <div
      class="_metadata_0a2670"
    >
      <span
        class="_msgNum_0a2670"
      >
        #
        1
      </span>
      <span
        class="_username--M_0a2670"
      >
        TestUser
      </span>
      <span
        class="_replyTime_0a2670"
      >
        2021年6月30日 19:00:00
      </span>
    </div>
    <div
      class="_msg_0a2670"
    >
      This is a test message
    </div>
    <div
      class="_votes_0a2670"
    >
      <div
        class="MuiStack-root css-so8pc8-MuiStack-root"
      >
        <div
          class="MuiStack-root css-1rlitzi-MuiStack-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="ThumbUpOffAltIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m13.11 5.72-.57 2.89c-.12.59.04 1.2.42 1.66.38.46.94.73 1.54.73H20v1.08L17.43 18H9.34c-.18 0-.34-.16-.34-.34V9.82zM14 2 7.59 8.41C7.21 8.79 7 9.3 7 9.83v7.83C7 18.95 8.05 20 9.34 20h8.1c.71 0 1.36-.37 1.72-.97l2.67-6.15c.11-.25.17-.52.17-.8V11c0-1.1-.9-2-2-2h-5.5l.92-4.65c.05-.22.02-.46-.08-.66-.23-.45-.52-.86-.88-1.22zM4 9H2v11h2c.55 0 1-.45 1-1v-9c0-.55-.45-1-1-1"
            />
          </svg>
           
          10
        </div>
        <div
          class="MuiStack-root css-1rlitzi-MuiStack-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="ThumbDownOffAltIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m10.89 18.28.57-2.89c.12-.59-.04-1.2-.42-1.66-.38-.46-.94-.73-1.54-.73H4v-1.08L6.57 6h8.09c.18 0 .34.16.34.34v7.84zM10 22l6.41-6.41c.38-.38.59-.89.59-1.42V6.34C17 5.05 15.95 4 14.66 4h-8.1c-.71 0-1.36.37-1.72.97l-2.67 6.15c-.11.25-.17.52-.17.8V13c0 1.1.9 2 2 2h5.5l-.92 4.65c-.05.22-.02.46.08.66.23.45.52.86.88 1.22zm10-7h2V4h-2c-.55 0-1 .45-1 1v9c0 .55.45 1 1 1"
            />
          </svg>
           
          2
        </div>
      </div>
    </div>
  </div>
</div>
`;
